### PR TITLE
librados: avoid symbol versioning on Windows

### PR DESCRIPTION
**Pull request by petrutlucian94**
Avoid symbol versioning when targeting Windows

symver isn't available on Windows. For now, we'll just pick the
default function version.

In the future, we may want to reconsider the versioning approach.

Signed-off-by: Lucian Petrut <lpetrut@cloudbasesolutions.com>

### CODE REVIEW

**Comment by wjwithagen** [comment]
@petrutlucian94 
You at least have to prefix the commit with something like `librados:` 
And perhaps add the `win32` to the end of the commit header.

**Comment by petrutlucian94** [comment]
> @petrutlucian94
> You at least have to prefix the commit with something like `librados:`
> And perhaps add the `win32` to the end of the commit header.

Are you referring to a specific commit, all of them (some are not strictly tied to librados) or the PR? That "[win32]" in the title makes it more visible, easier to understand what's it about or search for related commits.

File: src/librados/librados_c.cc [hunk 1]
```diff
@@ -41,13 +41,32 @@
...
 #define LIBRADOS_C_API_BASE(fn)               \
   asm(".symver _" #fn "_base, " #fn "@")
 #define LIBRADOS_C_API_BASE_DEFAULT(fn)       \
   asm(".symver _" #fn ", " #fn "@@")
 #define LIBRADOS_C_API_DEFAULT(fn, ver)       \
   asm(".symver _" #fn ", " #fn "@@LIBRADOS_" #ver)
 
+#define LIBRADOS_C_API_BASE_F(fn) _ ## fn ## _base
+#define LIBRADOS_C_API_BASE_DEFAULT_F(fn) _ ## fn
+#define LIBRADOS_C_API_DEFAULT_F(fn, ver) _ ## fn
```
Comment (tchaikov) [inline src/librados/librados_c.cc]: if the version number is always dropped on the floor, why bother passing it?

Comment (tchaikov) [reply src/librados/librados_c.cc]: and why do we need to define macros for creating the function names?

Comment (petrutlucian94) [reply src/librados/librados_c.cc]: > if the version number is always dropped on the floor, why bother passing it?

Right, that's a leftover. Initially, I was thinking about appending the version to the function name.

> and why do we need to define macros for creating the function names?

On Linux, an underscore will be prepended to the function names and the right one will be picked through symbol versioning. This isnt't supported on Windows.

**Comment by petrutlucian94** [comment]
> It would be nice if perhaps somehow you could generate an DEF export table and just alias the exported symbols to the proper names so that you don't need to wrap each function name w/ a macro. Do you have any expected users of the C API? librbd basically only uses the C++ API internally, so perhaps you really only need to expose a few methods.

We were hoping to support the Python bindings at some point, as well as FFI bindings for other languages, which most probably require the C API. At the same time, librados_cxx.cc itself uses functions from librados_c.cc [1].

At the moment, we have to use static libraries because some ceph components exceed the number of symbols allowed by a DLL (e.g. due to template usage), so we can't use import library hacks to rename symbols. Plus, we'd have issues at compile time.

[1] http://paste.openstack.org/raw/792949/

File: src/librados/librados_c.cc [hunk 2]
```diff
@@ -122,7 +141,7 @@ LIBRADOS_C_API_BASE_DEFAULT(rados_create);
 // 3) flags is for future expansion (maybe some of the global_init()
 //    behavior is appropriate for some consumers of librados, for instance)
 
-extern "C" int _rados_create2(rados_t *pcluster, const char *const clustername,
+extern "C" int LIBRADOS_C_API_BASE_DEFAULT_F(rados_create2)(rados_t *pcluster, const char *const clustername,
```
[the remark this answers is not in the document]
Comment (petrutlucian94) [reply src/librados/librados_c.cc:144]: Indeed, we can address that if we're to go with this approach.

File: src/librados/librados_c.cc [hunk 3]
```diff
@@ -41,13 +41,32 @@
...
+#else
+// symver cannot be used on Windows. We'll only be able
+// to support one version, unless we use a different
+// versioning approach.
+#define LIBRADOS_C_API_BASE(fn)
+#define LIBRADOS_C_API_BASE_DEFAULT(fn)
+#define LIBRADOS_C_API_DEFAULT(fn, ver)
+
+#define LIBRADOS_C_API_BASE_F(fn) _ ## fn ## _base
+#define LIBRADOS_C_API_BASE_DEFAULT_F(fn) fn
```
[the remark this answers is not in the document]
Comment (petrutlucian94) [reply src/librados/librados_c.cc:64]: Between ``LIBRADOS_C_API_BASE_F`` and ``LIBRADOS_C_API_BASE_DEFAULT_F``? The former uses the ``_base suffix`` and "_" prefix.

[the remark this answers is not in the document]
Comment (dillaman) [reply src/librados/librados_c.cc:64]: No, `LIBRADOS_C_API_BASE_DEFAULT_F` and `LIBRADOS_C_API_DEFAULT_F`

**Comment by dillaman** [comment]
> At the same time, librados_cxx.cc itself uses functions from librados_c.cc [1].

That shouldn't really matter since it could use an internal / private version.

> At the moment, we have to use static libraries because some ceph components exceed the number of symbols allowed by a DLL (e.g. due to template usage), so we can't use import library hacks to rename symbols. 

Really? The linker barfs on too-many private (non-exported) symbols? 

> I'm a bit concerned about symbol versioning since this isn't a portable way of tackling API versioning. 

Portable, no -- but it surely beats the alternative of having to (1) release a new libradosXYZ major with each Ceph release and 100% break backward compatibility with integrated products like QEMU, or (2) never allowing the API to change.

File: src/librados/librados_c.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (petrutlucian94) [reply src/librados/librados_c.cc:64]: Oh, right. Those can definitely be combined.

**Comment by petrutlucian94** [comment]
> That shouldn't really matter since it could use an internal / private version.

That works as well. If we decide that the current approach is unacceptable and that we don't need to support FFI, I can switch to that. That would imply moving those functions to a separate file and ensuring that librados_cxx.cc doesn't rely on librados_c.cc.

> Really? The linker barfs on too-many private (non-exported) symbols?

Yep, that also happens when using "-O0". We've tried addressing a few situations and we'll probably revisit this topic soon.

**Comment by wjwithagen** [comment]
> > I'm a bit concerned about symbol versioning since this isn't a portable way of tackling API versioning.
> 
> Portable, no -- but it surely beats the alternative of having to (1) release a new libradosXYZ major with each Ceph release and 100% break backward compatibility with integrated products like QEMU, or (2) never allowing the API to change.

@dillaman 
I understand the motivation of doing this.
But thusfar I have not been able to get this working with the linker that comes with the LLVM project. I just does not want to understand the linkermaps the way GNU-ld does.
And the consequence from that is that I need to use the GNU binutils linker to get this part of the code going.

So I expect that there will be more platform that will have the same type of problems, IF they do not use GNU-ld.

**Comment by smithfarm** [comment]
@petrutlucian94 Needs rebase

**Comment by petrutlucian94** [comment]
jenkins test this please

**Comment by petrutlucian94** [comment]
jenkins test this please

**Comment by petrutlucian94** [comment]
jenkins test api

**Comment by jan--f** [comment]
@dillaman @neha-ojha may I kindly request one last round of reviews for the windows rados related parts please?

**Review by jdurgin — APPROVED** [review]
this is a straight-forward approach, seems good for now

**Comment by jan--f** [comment]
Test results start to trickle in: http://pulpito.ceph.com/jafaj-2021-01-12_18:34:36-rados-wip-jan-testing-2021-01-12-1728-distro-basic-smithi/

**Comment by neha-ojha** [comment]
> Test results start to trickle in: http://pulpito.ceph.com/jafaj-2021-01-12_18:34:36-rados-wip-jan-testing-2021-01-12-1728-distro-basic-smithi/

I would like to merge this today but looks like tests are stuck

**Comment by petrutlucian94** [comment]
@neha-ojha @jan--f We now have the test results. I'll send a backport as soon as it's merged.

**Comment by neha-ojha** [comment]
> @neha-ojha @jan--f We now have the test results. I'll send a backport as soon as it's merged.

I will take a look today.

**Comment by neha-ojha** [comment]
> > @neha-ojha @jan--f We now have the test results. I'll send a backport as soon as it's merged.
> 
> I will take a look today.

@petrutlucian94 @jan--f Several failures seen in https://pulpito.ceph.com/jafaj-2021-01-12_18:34:36-rados-wip-jan-testing-2021-01-12-1728-distro-basic-smithi/ have already been addressed in latest master. I will run this PR on top of latest master to get a cleaner run and report back.

**Comment by neha-ojha** [comment]
https://pulpito.ceph.com/nojha-2021-01-20_23:46:47-rados-wip-32027-distro-basic-smithi/ - no related failures

**Pull request merged**